File: morloc-manager/src/MorlocManager/Freeze.hs
{- |
Module      : MorlocManager.Freeze
Description : Export installed morloc state as a portable frozen artifact
Copyright   : (c) Zebulun Arendsee, 2016-2026
License     : Apache-2.0
Maintainer  : z@morloc.io

Captures the complete installed morloc state (modules, programs, pools) as
a tarball with a JSON manifest, suitable for building an immutable serve
image.

@
  Builder Container           Frozen Artifact           Serve Image
  +------------------+       +------------------+     +------------------+
  | lib/  (modules)  |       | state.tar.gz     |     | read-only        |
  | fdb/  (manifests)| tar   |   lib/           |COPY | fdb + lib        |
  | bin/  (programs) |------>|   fdb/           |---->| bin + pools      |
  | pool files       |       |   bin/           |     | nexus router     |
  | morloc binary    |       |   morloc (bin)   |     | morloc (eval)    |
  +------------------+       | manifest.json    |     +------------------+
                              +------------------+
  FROZEN (excluded):
    GHC, stack, Cabal, build tools, package caches
@

The freeze manifest (@freeze-manifest.json@) provides an auditable record:
morloc version, installed modules with SHA-256 checksums, compiled programs
with their exported commands, base image reference, and environment layer.
This manifest can be reviewed before deploying to production.
-}

module MorlocManager.Freeze
  ( -- * Freeze operation
    freeze
    -- * Manifest
  , writeFreezeManifest
  , readFreezeManifest
  ) where

import qualified Crypto.Hash.SHA256 as SHA256
import qualified Data.Aeson as Aeson
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import Data.Text (Text)
import qualified Data.Text as Text
import Data.Time (getCurrentTime)
import Data.Word (Word8)
import System.Directory (doesDirectoryExist, doesFileExist, listDirectory, createDirectoryIfMissing)
import System.Exit (ExitCode(..))
import System.FilePath ((</>))
import System.IO (hPutStrLn, hFlush, stderr)
import System.Process (readProcessWithExitCode)

import MorlocManager.Types
import MorlocManager.Config (readVersionConfig, readActiveConfig, versionDataDir)

-- | Freeze the installed morloc state into a tarball and manifest.
--
-- The output directory receives:
--
--   * @state.tar.gz@ -- compressed archive of lib\/, fdb\/, bin\/, and the
--     morloc compiler binary
--   * @freeze-manifest.json@ -- auditable record of frozen contents
--
-- This runs inside (or against the data volume of) the builder container.
-- It does NOT freeze GHC, stack, or other build tooling.
freeze :: Scope -> Version -> FilePath -> IO (Either ManagerError ())
freeze scope ver outputDir = do
  createDirectoryIfMissing True outputDir
  vDataDir <- versionDataDir scope ver
  exists <- doesDirectoryExist vDataDir
  if not exists
    then pure (Left (VersionNotInstalled ver))
    else do
      -- Build the tarball
      hPutStrLn stderr $ "Freezing installed state from " <> vDataDir <> "..."
      hFlush stderr
      let tarPath = outputDir </> "state.tar.gz"
      (code, _stdout, tarErr) <- readProcessWithExitCode "tar"
        [ "-czf", tarPath
        , "-C", vDataDir
        , "lib", "fdb", "bin"
        ] ""
      case code of
        ExitFailure _ -> pure (Left (FreezeError ("tar failed: " <> tarErr)))
        ExitSuccess -> do
          hPutStrLn stderr $ "Created " <> tarPath
          -- Scan installed modules and programs for manifest
          modules' <- scanModules (vDataDir </> "lib")
          programs' <- scanPrograms (vDataDir </> "fdb")
          now <- getCurrentTime
          -- Read version config for image reference and active environment
          vcResult <- readVersionConfig scope ver
          mCfg <- readActiveConfig
          let baseImg = case vcResult of
                Right vc -> vcImage vc
                Left _   -> "unknown"
              envLyr = case mCfg of
                Just cfg | configActiveEnv cfg /= "base" -> Just (configActiveEnv cfg)
                _ -> Nothing
          let manifest = FreezeManifest
                { fmMorlocVersion = ver
                , fmFrozenAt      = now
                , fmModules       = modules'
                , fmPrograms      = programs'
                , fmBaseImage     = baseImg
                , fmEnvLayer      = envLyr
                }
          let manifestPath = outputDir </> "freeze-manifest.json"
          writeFreezeManifest manifestPath manifest
          hPutStrLn stderr $ "Wrote " <> manifestPath
          hPutStrLn stderr $ "Frozen state written to " <> outputDir
          hFlush stderr
          pure (Right ())

-- | Write a freeze manifest as pretty-printed JSON.
writeFreezeManifest :: FilePath -> FreezeManifest -> IO ()
writeFreezeManifest path manifest =
  BL.writeFile path (Aeson.encode manifest)

-- | Read a freeze manifest from a JSON file.
readFreezeManifest :: FilePath -> IO (Either ManagerError FreezeManifest)
readFreezeManifest path = do
  bytes <- BL.readFile path
  case Aeson.eitherDecode bytes of
    Left msg  -> pure (Left (FreezeError ("Invalid manifest: " <> msg)))
    Right val -> pure (Right val)

-- ======================================================================
-- Internal: scanning installed state
-- ======================================================================

-- | Scan the lib/ directory for installed modules.
--
-- For each subdirectory in lib/, reads the module name from the directory
-- name and computes a SHA-256 hash of all .loc files concatenated.
scanModules :: FilePath -> IO [ModuleEntry]
scanModules libDir = do
  exists <- doesDirectoryExist libDir
  if not exists
    then pure []
    else do
      entries <- listDirectory libDir
      dirs <- filterM (doesDirectoryExist . (libDir </>)) entries
      mapM (scanOneModule libDir) dirs

-- | Scan a single module directory for its name, version, and content hash.
scanOneModule :: FilePath -> String -> IO ModuleEntry
scanOneModule libDir name = do
  let modDir = libDir </> name
  -- Hash all .loc files in the module directory for a content fingerprint
  files <- listDirectory modDir
  let locFiles = [f | f <- files, hasSuffix ".loc" f]
  contentHash <- hashDirectoryFiles modDir locFiles
  pure ModuleEntry
    { meName    = Text.pack name
    , meVersion = Nothing  -- modules don't yet have standardized version metadata
    , meSha256  = contentHash
    }

-- | Scan the fdb/ directory for compiled program manifests.
--
-- For each @.manifest@ file, extracts the program name and parses the
-- JSON to find exported command names.
scanPrograms :: FilePath -> IO [ProgramEntry]
scanPrograms fdbDir = do
  exists <- doesDirectoryExist fdbDir
  if not exists
    then pure []
    else do
      entries <- listDirectory fdbDir
      let manifests = [e | e <- entries, hasSuffix ".manifest" e]
      mapM (scanOneProgram fdbDir) manifests

-- | Parse a single .manifest JSON file for program name and commands.
scanOneProgram :: FilePath -> String -> IO ProgramEntry
scanOneProgram fdbDir filename = do
  let path = fdbDir </> filename
      progName = stripSuffix ".manifest" filename
  commands <- parseManifestCommands path
  pure ProgramEntry
    { peName     = Text.pack progName
    , peCommands = commands
    }

-- | Extract command names from a manifest JSON file.
--
-- Parses the manifest as a simple structure with a @"commands"@ array,
-- each element having a @"name"@ field. Returns an empty list if the
-- file cannot be parsed.
parseManifestCommands :: FilePath -> IO [Text]
parseManifestCommands path = do
  exists <- doesFileExist path
  if not exists
    then pure []
    else do
      bytes <- BL.readFile path
      case Aeson.decode bytes :: Maybe ManifestStub of
        Nothing   -> pure []
        Just stub -> pure (map mscName (msCommands stub))

-- | Minimal manifest structure for extracting command names during freeze.
-- We only need the command names, not the full manifest schema.
data ManifestStub = ManifestStub
  { msCommands :: [ManifestStubCmd]
  } deriving (Show)

data ManifestStubCmd = ManifestStubCmd
  { mscName :: Text
  } deriving (Show)

instance Aeson.FromJSON ManifestStub where
  parseJSON = Aeson.withObject "ManifestStub" $ \o ->
    ManifestStub <$> o Aeson..:? "commands" Aeson..!= []

instance Aeson.FromJSON ManifestStubCmd where
  parseJSON = Aeson.withObject "ManifestStubCmd" $ \o ->
    ManifestStubCmd <$> o Aeson..: "name"

-- ======================================================================
-- Internal utilities
-- ======================================================================

-- | Compute SHA-256 hash of concatenated file contents from a directory.
hashDirectoryFiles :: FilePath -> [String] -> IO Text
hashDirectoryFiles dir files = do
  contents <- mapM (\f -> BS.readFile (dir </> f)) (sort files)
  let digest = SHA256.hash (BS.concat contents)
  pure (hexEncode digest)

-- | Encode a ByteString as a hex Text.
hexEncode :: BS.ByteString -> Text
hexEncode bs = Text.pack (concatMap byteToHex (BS.unpack bs))
  where
    byteToHex :: Word8 -> String
    byteToHex b =
      let (hi, lo) = b `divMod` 16
      in [hexChar hi, hexChar lo]
    hexChar :: Word8 -> Char
    hexChar n
      | n < 10    = toEnum (fromEnum '0' + fromIntegral n)
      | otherwise = toEnum (fromEnum 'a' + fromIntegral n - 10)

hasSuffix :: String -> String -> Bool
hasSuffix suffix str = drop (length str - length suffix) str == suffix

stripSuffix :: String -> String -> String
stripSuffix suffix str
  | hasSuffix suffix str = take (length str - length suffix) str
  | otherwise = str

sort :: Ord a => [a] -> [a]
sort [] = []
sort (x:xs) = sort [y | y <- xs, y <= x] ++ [x] ++ sort [y | y <- xs, y > x]

filterM :: Monad m => (a -> m Bool) -> [a] -> m [a]
filterM _ [] = pure []
filterM p (x:xs) = do
  ok <- p x
  rest <- filterM p xs
  pure (if ok then x : rest else rest)

File: data/rust/morloc-manager/src/freeze.rs
use std::fs;
use std::path::Path;
use std::process::{Command, Stdio};

use chrono::Utc;
use sha2::{Digest, Sha256};
use crate::config;
use crate::error::{ManagerError, Result};
use crate::types::*;

pub fn freeze_from_dir(
    scope: Scope,
    ver: Version,
    engine: ContainerEngine,
    image: &str,
    v_data_dir: &str,
    output_dir: &str,
    verbose: bool,
    env_vars: &[String],
) -> Result<()> {
    fs::create_dir_all(output_dir)
        .map_err(|e| ManagerError::FreezeError(format!("Failed to create output dir: {e}")))?;

    if !Path::new(v_data_dir).is_dir() {
        return Err(ManagerError::FreezeError(format!(
            "Data directory does not exist: {v_data_dir}"
        )));
    }

    // Validate programs exist before writing any files
    let modules = scan_modules(&format!("{v_data_dir}/fdb"));
    let programs = scan_programs(&format!("{v_data_dir}/fdb"));
    if programs.is_empty() {
        return Err(ManagerError::FreezeError(
            "No morloc programs are installed. Compile and install with 'morloc make --install' before freezing.".to_string()
        ));
    }

    // Validate programs work before freezing
    let mh = "/opt/morloc";
    let bind_mounts = vec![(v_data_dir.to_string(), mh.to_string())];
    crate::serve::validate_programs(engine, image, &programs, bind_mounts, verbose)?;

    eprintln!("Freezing installed state from {v_data_dir}...");
    let tar_path = Path::new(output_dir).join("state.tar.gz");
    let tar_path = tar_path.to_string_lossy();
    let mut tar_dirs: Vec<&str> = Vec::new();
    for dir in &["lib", "fdb", "bin", "exe", "opt", "src"] {
        if Path::new(&format!("{v_data_dir}/{dir}")).is_dir() {
            tar_dirs.push(dir);
        }
    }
    let tar_status = Command::new("tar")
        .args(["-czf", &tar_path, "-C", v_data_dir])
        .args(&tar_dirs)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::inherit())
        .status()
        .map_err(|e| ManagerError::FreezeError(format!("tar failed: {e}")))?;

    if !tar_status.success() {
        return Err(ManagerError::FreezeError(
            "tar failed (see error output above)".to_string()
        ));
    }
    eprintln!("Created {tar_path}");
    let now = Utc::now();

    // Get base image from the active environment config.
    // Check local config first, then system config for the active env name
    // (mirrors resolve_active_env_name in environment.rs).
    let active_env_name: Option<String> = config::read_active_config()
        .and_then(|c| c.active_env)
        .or_else(|| {
            let sys_path = config::config_path(Scope::System);
            config::read_config::<Config>(&sys_path)
                .ok()
                .and_then(|c| c.active_env)
        });

    let (base_img, env_layer) = if let Some(ref env_name) = active_env_name {
        let env_scope = config::find_env_scope(env_name).unwrap_or(scope);
        match config::read_env_config(env_scope, env_name) {
            Ok(ec) => {
                let base = ec.base_image.clone();
                // Capture env layer info if there's a Dockerfile
                let layer = if ec.dockerfile.is_some() {
                    let df_path = config::env_dockerfile_path(env_scope, env_name);
                    if df_path.exists() {
                        let df_contents = fs::read_to_string(&df_path).unwrap_or_default();
                        let content_hash = ec.content_hash.unwrap_or_default();
                        // Use the tagged image reference (not digest) so that
                        // unfreeze can resolve it locally without network access.
                        // Digest references like localhost/morloc-env@sha256:...
                        // cause BuildKit to attempt HTTPS to localhost.
                        let image_tag = ec.built_image.clone();
                        Some(FrozenEnvLayer {
                            name: env_name.to_string(),
                            dockerfile: df_contents,
                            content_hash,
                            image_tag,
                        })
                    } else {
                        None
                    }
                } else {
                    None
                };
                (base, layer)
            }
            Err(_) => ("unknown".to_string(), None),
        }
    } else {
        ("unknown".to_string(), None)
    };

    let manifest = FreezeManifest {
        morloc_version: ver,
        frozen_at: now,
        modules,
        programs,
        base_image: base_img,
        env_layer,
        env_vars: env_vars.to_vec(),
    };
    let manifest_path = Path::new(output_dir).join("freeze-manifest.json");
    let manifest_path = manifest_path.to_string_lossy();
    write_freeze_manifest(&manifest_path, &manifest)?;
    eprintln!("Wrote {manifest_path}");
    if !env_vars.is_empty() {
        eprintln!("Expected env vars: {}", env_vars.join(", "));
        eprintln!("  Pass at start time with: morloc-manager start --env KEY=VALUE");
    }
    eprintln!("Frozen state written to {output_dir}");
    Ok(())
}

pub fn write_freeze_manifest(path: &str, manifest: &FreezeManifest) -> Result<()> {
    let json = serde_json::to_vec(manifest)
        .map_err(|e| ManagerError::FreezeError(format!("JSON encode failed: {e}")))?;
    fs::write(path, json)
        .map_err(|e| ManagerError::FreezeError(format!("Write failed: {e}")))?;
    Ok(())
}

pub fn read_freeze_manifest(path: &str) -> Result<FreezeManifest> {
    let bytes =
        fs::read(path).map_err(|e| ManagerError::FreezeError(format!("Read failed: {e}")))?;
    serde_json::from_slice(&bytes)
        .map_err(|e| ManagerError::FreezeError(format!("Invalid manifest: {e}")))
}

// ======================================================================
// Internal: scanning installed state
// ======================================================================

fn scan_modules(fdb_dir: &str) -> Vec<ModuleEntry> {
    let fdb_path = Path::new(fdb_dir);
    if !fdb_path.is_dir() {
        return Vec::new();
    }
    let Ok(entries) = fs::read_dir(fdb_path) else {
        return Vec::new();
    };

    #[derive(serde::Deserialize)]
    struct ModuleStub {
        name: String,
        #[serde(default)]
        version: Option<String>,
    }

    entries
        .flatten()
        .filter(|e| {
            e.file_name()
                .to_string_lossy()
                .ends_with(".module")
        })
        .filter_map(|e| {
            let bytes = fs::read(e.path()).ok()?;
            let stub: ModuleStub = serde_json::from_slice(&bytes).ok()?;
            let digest = Sha256::digest(&bytes);
            let sha256: String = digest.iter().map(|b| format!("{b:02x}")).collect();
            Some(ModuleEntry {
                name: stub.name,
                version: stub.version,
                sha256,
            })
        })
        .collect()
}

fn scan_programs(fdb_dir: &str) -> Vec<ProgramEntry> {
    let fdb_path = Path::new(fdb_dir);
    if !fdb_path.is_dir() {
        return Vec::new();
    }
    let Ok(entries) = fs::read_dir(fdb_path) else {
        return Vec::new();
    };
    entries
        .flatten()
        .filter(|e| {
            e.file_name()
                .to_string_lossy()
                .ends_with(".manifest")
        })
        .map(|e| {
            let filename = e.file_name().to_string_lossy().to_string();
            let prog_name = filename.strip_suffix(".manifest").unwrap_or(&filename);
            let commands = parse_manifest_commands(&e.path());
            ProgramEntry {
                name: prog_name.to_string(),
                commands,
            }
        })
        .collect()
}

fn parse_manifest_commands(path: &Path) -> Vec<String> {
    let Ok(bytes) = fs::read(path) else {
        return Vec::new();
    };
    #[derive(serde::Deserialize)]
    struct ManifestStub {
        #[serde(default)]
        commands: Vec<ManifestStubCmd>,
    }
    #[derive(serde::Deserialize)]
    struct ManifestStubCmd {
        name: String,
    }
    match serde_json::from_slice::<ManifestStub>(&bytes) {
        Ok(stub) => stub.commands.into_iter().map(|c| c.name).collect(),
        Err(_) => Vec::new(),
    }
}
